src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs

```diff
@@ -86,52 +88,65 @@ private static unsafe ulong GetMaskAsLong(byte[] bytes)
         }
 
         // The same as GetAsciiStringNonNullCharacters but throws BadRequest
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static unsafe string GetHeaderName(this ReadOnlySpan<byte> span)
         {
             if (span.IsEmpty)
             {
                 return string.Empty;
             }
 
-            var asciiString = new string('\0', span.Length);
+            fixed (byte* source = &MemoryMarshal.GetReference(span))
+            {
+                return string.Create(span.Length, new IntPtr(source), s_getHeaderName);
+            }
+        }
 
-            fixed (char* output = asciiString)
-            fixed (byte* buffer = span)
+        private static readonly SpanAction<char, IntPtr> s_getHeaderName = GetHeaderName;
+
+        private static unsafe void GetHeaderName(Span<char> buffer, IntPtr state)
+        {
+            fixed (char* output = &MemoryMarshal.GetReference(buffer))
             {
                 // This version if AsciiUtilities returns null if there are any null (0 byte) characters
                 // in the string
-                if (!StringUtilities.TryGetAsciiString(buffer, output, span.Length))
+                if (!StringUtilities.TryGetAsciiString((byte*)state.ToPointer(), output, buffer.Length))
                 {
                     BadHttpRequestException.Throw(RequestRejectionReason.InvalidCharactersInHeaderName);
                 }
             }
-
-            return asciiString;
         }
 
         public static string GetAsciiStringNonNullCharacters(this Span<byte> span)
             => GetAsciiStringNonNullCharacters((ReadOnlySpan<byte>)span);
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static unsafe string GetAsciiStringNonNullCharacters(this ReadOnlySpan<byte> span)
         {
             if (span.IsEmpty)
             {
                 return string.Empty;
             }
 
-            var asciiString = new string('\0', span.Length);
+            fixed (byte* source = &MemoryMarshal.GetReference(span))
+            {
+                return string.Create(span.Length, new IntPtr(source), s_getAsciiStringNonNullCharacters);
+            }
+        }
 
-            fixed (char* output = asciiString)
-            fixed (byte* buffer = span)
+        private static readonly SpanAction<char, IntPtr> s_getAsciiStringNonNullCharacters = GetAsciiStringNonNullCharacters;
```

Comment:
ditto for s_getHeaderName

Reply:
Done -- will be updated with the next push.